src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/ActiveDirectoryAuthenticationProvider.cs

```diff
@@ -284,14 +298,37 @@ previousPw is byte[] previousPwBytes &&
                     // An 'MsalUiRequiredException' is thrown in the case where an interaction is required with the end user of the application,
                     // for instance, if no refresh token was in the cache, or the user needs to consent, or re-sign-in (for instance if the password expired),
                     // or the user needs to perform two factor authentication.
-                    result = await AcquireTokenInteractiveDeviceFlowAsync(app, scopes, parameters.ConnectionId, parameters.UserId, parameters.AuthenticationMethod, cts, _customWebUI, _deviceCodeFlowCallback).ConfigureAwait(false);
+
+#if NETFRAMEWORK
+                    Func<Task<AuthenticationResult>> func = async () =>
+                    {
+                        return await AcquireTokenInteractiveDeviceFlowAsync(app, scopes, parameters.ConnectionId, parameters.UserId, parameters.AuthenticationMethod, cts, _customWebUI, _deviceCodeFlowCallback
+                        ).ConfigureAwait(false);
+                    };
+
+                    result = await (Task<AuthenticationResult>)_control.Invoke(func);
```

Comment:
Similar to the above - but we only really need a [`SynchronizationContext`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.synchronizationcontext?view=net-8.0&viewFallbackFrom=windowsdesktop-8.0) here. We can pass in an instance of [`WindowsFormsSynchronizationContext`](https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.windowsformssynchronizationcontext?view=windowsdesktop-8.0) or [`DispatcherSynchronizationContext`](https://learn.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatchersynchronizationcontext?view=windowsdesktop-8.0&viewFallbackFrom=net-8.0), which enables cross-target WPF support.

SynchronizationContext.Send is a little awkward to use - we can pass in a state object and set a property on it, but we don't get a return value.

I'm pretty sure that this (and the almost-identical code below) eliminates the System.Windows.Forms dependency.

Reply:
This makes sense to me, but I'll want to think about how best to expose this in the public API. The nice thing about passing in a Control is that it gives parent window and dispatcher together in one object. Having a separate SynchronizationContext specifiable will require the API to be clear that it's required when using the parent window param.

Reply:
I see what you're talking about! When using SynchronizationContext.Send, I think we'd be hedged in to either the semaphore approach you've taken, or a TaskCompletionSource - but it'll have the same net effect.

Would it be workable to record `SynchronizationContext.Current`, call `SynchronizationContext.SetSynchronizationContext` to force the thread into the right sync context, await AcquireTokenInteractive, then revert back to the original sync context in a `finally` block? I'm less sure, but I suspect that this await on AcquireTokenInteractive might actually need us to pass true to `ConfigureAwait` (so it'll continue on the original sync context, ready for the `finally` block to revert it.)

Reply:
I tried this out and it didn't give me the behavior I expected. It started returning cross-thread access exceptions again. I think the catch is that setting the synchronization context doesn't actually switch you to another thread, it only affects where the continuations run. So the first thing that actually runs async still runs on "this" thread, which is not guaranteed to be the UI thread.

So this doesn't work:
```c#
var currentSynchronizationContext = SynchronizationContext.Current;

try
{
    SynchronizationContext.SetSynchronizationContext(_synchronizationContext);
    result = await AcquireTokenInteractiveDeviceFlowAsync(app, scopes, parameters.ConnectionId, parameters.UserId, parameters.AuthenticationMethod, cts, _customWebUI, _deviceCodeFlowCallback);
}
finally
{
    SynchronizationContext.SetSynchronizationContext(currentSynchronizationContext);
}
```

But something like this does:
```c#
var currentSynchronizationContext = SynchronizationContext.Current;

try
{
    SynchronizationContext.SetSynchronizationContext(_synchronizationContext);

    var t = Task.Delay(500);

    result = await await t.ContinueWith(antecedent =>

         AcquireTokenInteractiveDeviceFlowAsync(app, scopes, parameters.ConnectionId, parameters.UserId, parameters.AuthenticationMethod, cts, _customWebUI, _deviceCodeFlowCallback)

        , TaskScheduler.FromCurrentSynchronizationContext());

}
finally
{
    SynchronizationContext.SetSynchronizationContext(currentSynchronizationContext);
}
```

Reply:
But I also came across this solution that may be cleaner: https://devblogs.microsoft.com/pfxteam/implementing-a-synchronizationcontext-sendasync-method/

Reply:
The above didn't work exactly, but led to a similar approach using TaskCompletionSource that I think it more readable than manually switching between synchronization contexts.

Reply:
Definitely, thanks - that looks a lot neater than my suggestion!

The only edge case I can think of is that ASP.NET Core doesn't have a synchronisation context at all. We'd hit a NullReferenceException if someone was trying to use an authentication method of ActiveDirectoryInteractive or ActiveDirectoryDeviceCodeFlow, which sounds like a misconfiguration to me. If we can't get a token and `_synchronizationContext` is null, it might be worth logging a trace event and throwing a more descriptive exception before trying to post to it.

Does interactive authentication now work with WPF on .NET Core? This scenario should test both of the situations I'm thinking of (any async changes between .NET Framework and Core, and any differences in synchronization context and cross-thread access between WPF and WinForms.)

Reply:
We can detect a default parent window for console apps using the pinvoke methods listed at the bottom of the changeset. When we use that, we don't need to execute the token acquisition in any particular synchronization context. I need to do another pass to incorporate that in a way that makes sense. It will also need to be included conditionally with the broker library, so I'm just holding off until I figure out how I want to do that.

I'll check out how things look with WPF, I've been testing mainly with winforms.

Reply:
Great! I think this gives us the best of both worlds then, until we're ready for MDS.Azure.Auth: we don't introduce any direct dependency on System.Windows.Forms in the .NET Core build (so although we've still got a transitive dependency when compiled for Windows, that'll disappear on the library split), and we expand functionality slightly, (to WPF and iOS & Android) since .NET Core can now call WithParentActivityOrWindow. Thanks for bearing with me on this.

Looking at the PInvokes, I've not got any comments on the overall idea. I have a few comments on some behaviour, but I'll drop them in the right place.